Added microphone ducking documentation

deborahtrez opened the pull request:
With reference to this issue #57 , i added a section describing microphone ducking, how it works, and the difference between manual and auto mode. 
@ywwg @Holzhaus @ronso0 please read through and let me know where i need to make changes ... :)

ronso0 commented:
Thanks! Will check this later on.

Just a small hint: when reviewing the commit history of the manual/any branch later it helps to have distinct branch names, like `deborahtrez/micducking` insead of `deborahtrez/patch-123`.

deborahtrez commented:
> Thanks! Will check this later on.
> 
> Just a small hint: when reviewing the commit history of the manual/any branch later it helps to have distinct branch names, like `deborahtrez/micducking` insead of `deborahtrez/patch-123`.

Okay. I have taken note of this. Thank you!

ywwg commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,27 @@ input.
+Clicking  on the :guilabel:`Auto` button will switch it to any of three modes:
+
+   * Off: When you talk through your microphone, the music  volume does not reduce.
+   * Auto: Talking through your microphone will automatically reduce the music volume. You can adjust the amount to which the volume is automatically reduced with the Strength knob.
```
"When you speak through the microphone, the music volume will be reduced depending on how loud you speak. You can adjust the amount by which the volume is automatically reduced with the Strength knob."

deborahtrez replied:
I have made the change. It sounds better :)

ywwg commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,27 @@ input.
+
+   * Off: When you talk through your microphone, the music  volume does not reduce.
+   * Auto: Talking through your microphone will automatically reduce the music volume. You can adjust the amount to which the volume is automatically reduced with the Strength knob.
+   * Manual: This sets a fixed level to which the music volume reduces to, regardless of the volume of the microphone input. When you talk through your microphone, you can manually set the amount to which the music volume is reduced by turning the Strength knob to a level that suits you.
```
I forget if the knob represents the volume is reduced *to*, or the amount the volume is reduced *by*.  I'm also not sure if those are identical things :).

ywwg replied:
"Manual: This sets a fixed level to which the music volume reduces to when the Talk button is on, regardless of the volume of the microphone input. When you talk through your microphone, you can manually set the amount to which the music volume is reduced by turning the Strength knob to a level that suits you."

I think that correctly describes how ducking works?  I honestly don't remember

deborahtrez replied:
I have edited Manual mode to differentiate it from the description. The new description teaches the user how to adjust the music volume when he talks, using the Strength knob in manual mode.

ywwg commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,27 @@ input.
+
+Microphone Ducking
+------------------
+While using Mixxx, you might want to talk through your microphone as the music plays, therefore, you might need the music to automatically fade into the background each time you speak, or you might prefer lowering the volume of the music to a specific level. Microphone ducking helps you achieve this.
```
The word "might" appears a lot here.  We can simplify:

"While using Mixxx with a microphone, you might want the music to automatically fade into the background each time you speak, or you might prefer lowering the volume of the music to a specific level. Microphone ducking helps you achieve this. The music will 'duck' out of the way of the microphone audio."

ywwg replied:
(I think it's worth explaining the word "duck" just in case people are having trouble with that)

deborahtrez replied:
Thank you. I agree. 
I have added a definition for ducking just before this paragraph on Line 417. You can read through and let me know if where it might need correction....

ywwg reviewed (changes requested):
I've added a few initial comments.  We will need to make sure this description aligns with how ducking actually works!  It's been so long I forget, and there have been bugs recently.

ronso0 commented:
> I've added a few initial comments. We will need to make sure this description aligns with how ducking actually works! It's been so long I forget, and there have been bugs recently.

...and keep in mind this targets the **2.2** manual!
@deborahtrez Do you mind double-checking how ducking behaves in Mixxx 2.2.4? Especially if the Strength knob behaves the same in both duck modes.
I remember we fixed a few things recently, but I _think_ those tweaks went into 2.3

deborahtrez commented:
> > I've added a few initial comments. We will need to make sure this description aligns with how ducking actually works! It's been so long I forget, and there have been bugs recently.
> 
> ...and keep in mind this targets the **2.2** manual!
> @deborahtrez Do you mind double-checking how ducking behaves in Mixxx 2.2.4? Especially if the Strength knob behaves the same in both duck modes.
> I remember we fixed a few things recently, but I _think_ those tweaks went into 2.3

Yes. I will do this and give you the report.

deborahtrez commented:
Hi @ronso0 . I have experimented with the Strength knob and this is what i have seen. When  you set the strength knob to a certain value and switch from Off to Auto,  without talking, the volume of the music will rise gradually from the ducking value set by the Strength knob,  to the actual music volume that is unaffected by ducking. Speaking through the microphone will then cause it to lower to the ducked value, but will keep rising back up in between talk sessions.

Switching to Manual mode..... The music volume will lower and play at the set ducked value, until it is manually changed by the user, by turning the Strength knob to whichever desired value. Even if the Strength knob is turned to the maximum (music here is at the loudest), and you speak through the mic, the volume of the music will not reduce, unless you turn the knob to reduce it to a volume loud enough to hear your own voice.

The level  by which the music volume reduces in both cases, is consistent. Meaning.... if i first set the Strength knob to a value that i wanted in Manual mode, then switched to Auto, the volume that the music is reduced to in Auto mode is the same as that in Manual mode.

The ducking effect behaves okay, in my opinion. The controls are working the way they should.

deborahtrez commented:
What can i do to fix the pre-commit (pull_request) check? @ronso0

Holzhaus commented:
Please run `git pull https://github.com/mixxxdj/mixxx.git manual-2.2.x`.

deborahtrez commented:
> Please run `git pull https://github.com/mixxxdj/mixxx.git manual-2.2.x`.

Okay. Let me do this. Thank you.

deborahtrez commented:
> Please run `git pull https://github.com/mixxxdj/mixxx.git manual-2.2.x`.

I run this but i got `Couldn't find remote ref manual-2.2.x` so i run `git pull https://github.com/mixxxdj/manual.git manual-2.2.` and it worked. Thanks @Holzhaus .

Holzhaus commented on source/chapters/microphones.rst:
```diff
@@ -430,9 +430,8 @@ The Strength knob adjusts the microphone talkover ducking strength and its behav
 
    * Off: When you speak through the microphone, the music  volume does not reduce.
-   * Auto: When you speak through the microphone, the music volume will be reduced when Mixxx detects your voice. You can adjust the amount by which the volume is automatically reduced with the Strength knob.
+   * Auto: When you speak through the microphone, the music volume will be reduced when Mixxx detects your voice. You can adjus4t the amount by which the volume is automatically reduced with the Strength knob.
```
This introduces a typo.

deborahtrez replied:
Ohhh my God. Let me get that fixed right now.

deborahtrez replied:
hmmm....check still didn't pass..

Holzhaus commented:
> I failed this check, but I'm not quite sure what to rectify. I was wondering if you might be able to help me identify the error so that i can make the correction....
> ![image](https://user-images.githubusercontent.com/28389950/97221953-a9919100-17de-11eb-9e83-0049a450e239.png)
> 

The newline byte at the end of the file is missing: https://stackoverflow.com/a/5813359

You need to configure your editor to add it. For example, if you're using Sublime Text you can set that option like this: https://coderwall.com/p/icxxna/don-t-forget-that-extra-line

deborahtrez commented:
Alright. Understood. I have configured  VSCode to automatically add the newline at the end of each file. Thank you! @Holzhaus ready for your review...

Holzhaus commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+
+Microphone Ducking
+------------------
+Microphone ducking is that effect you hear when you speak through the microphone, and the music volume is reduced by the presence of the voice signal coming through the microphone.
```
```suggestion
Microphone ducking is a technique to improve the audibility of speech by reducing the volume of background music while a voice signal is recorded by the microphone and allows listeners to hear the voice more clearly and consistently.
```

Holzhaus commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+
+   The microphone ducking controls
+
+The Strength knob adjusts the microphone talkover ducking strength and its behaviour depends on any of the Microphone talkover modes described below.
```
```suggestion
The :guilabel:`Strength` knob adjusts the microphone talkover ducking strength and its behaviour depends on any of the Microphone talkover modes described below.
```

ywwg replied:
@deborahtrez , you can incorporate jan's changes by selecting the "commit suggestion" button, or adding it to a batch.

deborahtrez replied:
Okay. I will start using this method. Thank you Owen.

Holzhaus commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+Microphone ducking is that effect you hear when you speak through the microphone, and the music volume is reduced by the presence of the voice signal coming through the microphone.
+While using Mixxx with a microphone, you might want the music to automatically fade into the background each time you speak, or you might prefer lowering the volume of the music to a specific level. Microphone ducking helps you achieve this.  The music will 'duck' out of the way of the microphone audio.
+
+.. figure:: ../_static/microphone_ducking.png
```
Can you put the screenshot in a `2.2` subdirectory and use the same naming scheme (dashes vs. underscores) like you do in #154?

deborahtrez replied:
Okay. I will.

Holzhaus commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+The Strength knob adjusts the microphone talkover ducking strength and its behaviour depends on any of the Microphone talkover modes described below.
+Clicking  on the :guilabel:`Auto` button will switch it to any of three modes:
+
+   * Off: When you speak through the microphone, the music  volume does not reduce.
+   * Auto: When you speak through the microphone, the music volume will be reduced when Mixxx detects your voice. You can adjust the amount by which the volume is automatically reduced with the Strength knob.
+   * Manual: In this mode, you have to remember to manually adjust the Strength knob to a desired level each time you speak. Otherwise, the music will play at the volume set by the Strength knob, regardless of the volume coming through the microphone input. When you speak through your microphone, set the volume to which the music should reduce by turning the Strength knob to a level that suits yo ...
```
```suggestion
   * :guilabel:`Off`: When you speak through the microphone, the music  volume does not reduce.
   * :guilabel:`Auto`: When you speak through the microphone, the music volume will be reduced when Mixxx detects your voice. You can adjust the amount by which the volume is automatically reduced with the Strength knob.
   * :guilabel:`Manual`: In this mode, you have to remember to manually adjust the Strength knob to a desired level each time you speak. Otherwise, the music will play at the volume set by the Strength knob, regardless of the volume coming through the microphone input. When you speak through your microphone, set the volume to which the music should reduce by turning the Strength knob to a level that suits you.
```

deborahtrez commented:
Hey @Holzhaus , I have implemented all the requested changes. Ready for your review....

ywwg commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+
+   * :guilabel:`Off`: When you speak through the microphone, the music  volume does not reduce.
+   * :guilabel:`Auto`: When you speak through the microphone, the music volume will be reduced when Mixxx detects your voice. You can adjust the amount by which the volume is automatically reduced with the Strength knob.
+   * :guilabel:`Manual`: In this mode, you have to remember to manually adjust the Strength knob to a desired level each time you speak. Otherwise, the music will play at the volume set by the Strength knob, regardless of the volume coming through the microphone input. When you speak through your microphone, set the volume to which the music should reduce by turning the Strength knob to a level t ...
```
this is true, but I think the intended behavior is for the user to use the Talk button to control this.

"In this mode, the music will play at the volume set by the Strength knob whenever the microphone Talk button is active, regardless of the volume coming through the microphone input. When you speak through your microphone, set the volume to which the music should reduce by turning the Strength knob to a level that suits you. When you disable Talk mode, the music will return to the regular volume"

deborahtrez replied:
Oh. Alright.

ywwg commented:
This is close to done! I made one more comment about Manual mode -- it's meant to be used in conjunction with the Talk button.  What you said is not wrong, but I wanted to clarify it.

deborahtrez commented:
> This is close to done! I made one more comment about Manual mode -- it's meant to be used in conjunction with the Talk button. What you said is not wrong, but I wanted to clarify it.

Alright. Let me change this, and commit. Thank you!

deborahtrez commented:
The linkcheck fail seems to be coming from the DJing_with_Mixx.rst file.  How are these broken links fixed?

Holzhaus commented:
> The linkcheck fail seems to be coming from the DJing_with_Mixx.rst file. How are these broken links fixed?

Don't worry, archive.org is somethimes slow to respond and the linkcheck times out after 5 seconds. If the link works locally, we can ignore it.

deborahtrez commented:
> > The linkcheck fail seems to be coming from the DJing_with_Mixx.rst file. How are these broken links fixed?
> 
> Don't worry, archive.org is somethimes slow to respond and the linkcheck times out after 5 seconds. If the link works locally, we can ignore it.

Okay then

Holzhaus commented:
@ywwg Please review, I'm not really familiar with mics in Mixxx.

ywwg commented on source/chapters/microphones.rst:
```diff
@@ -411,3 +411,28 @@ input.
+
+.. hint:: To set the Ducking volume correctly, you can use manual mode to adjust the knob to your taste and then set to auto when happy.
+
+.. Note:: Enabling the Microphone Gain option will cause the ducking effect to be ineffective, so, you can disable it for the ducking effect to work correctly.
```
@deborahtrez  where did this note come from? When you say "ineffective" do you mean "less effective", or "not effective at all?"

deborahtrez replied:
When I enabled the Microphone Gain, and used microphone ducking, I noticed that the two would not work together (at all). Even with very low and very high ducking Strengths. The Gain option had to be off/disabled for the ducking to work.

deborahtrez replied:
Is this note worthy?  Perhaps i should leave it out?

ywwg replied:
Yes that seems like a bug.  Leave it out and then this is LGTM

Holzhaus replied:
Please file a bug report on Launchpad then.

ronso0 replied:
> The Gain option had to be off/disabled for the ducking to wor

Are you talking about the tiny Gain 'mute' button in Deere 2.2?
When the input is muted there's no reason for the Ducking to kick in, so I'd say this is intended behaviour.

ronso0 replied:
Those gain buttons were removed for Mixxx 2.3 as they only cause confusion and we added PLAY buttons to Aux decks as equivalent for mics' TALK buttons for a consistent UI.

deborahtrez replied:
I have filed the report [here](https://bugs.launchpad.net/mixxx/+bug/1903754) on the launchpad

ronso0 replied:
Marked as Invalid because the source of confusion is that the GAIN label should describe the Gain knob only but is right next to the Mute button.
So you discovered a UX issue instead :)

ronso0 replied:
Repairing the UI in 2.2 would be the right fix, but I doubt the effort is worth it since we're rather close to publishing 2.3.

deborahtrez replied:
Okay. Let me remove this note section and commit.

deborahtrez commented:
Hey @ronso0 @ywwg @Holzhaus I've removed the note at the bottom, as requested. Kindly review...  :)

ywwg reviewed (approved):
LGTM.  Sorry this took so long!

deborahtrez commented:
> LGTM. Sorry this took so long!

No problem! Thank you! 😃
